internal/cloudapi/server.go

```diff
@@ -0,0 +1,226 @@
+//go:generate go run github.com/deepmap/oapi-codegen/cmd/oapi-codegen --package=cloudapi --generate types,chi-server,client -o openapi.gen.go openapi.yml
+
+package cloudapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/google/uuid"
+
+	"github.com/osbuild/osbuild-composer/internal/blueprint"
+	"github.com/osbuild/osbuild-composer/internal/distro"
+	"github.com/osbuild/osbuild-composer/internal/rpmmd"
+	"github.com/osbuild/osbuild-composer/internal/target"
+	"github.com/osbuild/osbuild-composer/internal/worker"
+)
+
+// Server represents the state of the cloud Server
+type Server struct {
+	workers     *worker.Server
+	rpmMetadata rpmmd.RPMMD
+	distros     *distro.Registry
+}
+
+// NewServer creates a new cloud server
+func NewServer(workers *worker.Server, rpmMetadata rpmmd.RPMMD, distros *distro.Registry) *Server {
+	server := &Server{
+		workers:     workers,
+		rpmMetadata: rpmMetadata,
+		distros:     distros,
+	}
+	return server
+}
+
+// Serve serves the cloud API over the provided listener socket
+func (server *Server) Serve(listener net.Listener) error {
+	s := http.Server{Handler: Handler(server)}
+
+	err := s.Serve(listener)
+	if err != nil && err != http.ErrServerClosed {
+		return err
+	}
+
+	return nil
+}
+
+// Compose handles a new /compose POST request
+func (server *Server) Compose(w http.ResponseWriter, r *http.Request) {
+	contentType := r.Header["Content-Type"]
+	if len(contentType) != 1 || contentType[0] != "application/json" {
+		http.Error(w, "Only 'application/json' content type is supported", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	var request ComposeRequest
+	err := json.NewDecoder(r.Body).Decode(&request)
+	if err != nil {
+		http.Error(w, "Could not parse JSON body", http.StatusBadRequest)
+		return
+	}
+
+	distribution := server.distros.GetDistro(request.Distribution)
+	if distribution == nil {
+		http.Error(w, fmt.Sprintf("Unsupported distribution: %s", request.Distribution), http.StatusBadRequest)
+		return
+	}
+
+	type imageRequest struct {
+		manifest distro.Manifest
+	}
+	imageRequests := make([]imageRequest, len(request.ImageRequests))
+	var targets []*target.Target
+
+	for i, ir := range request.ImageRequests {
+		arch, err := distribution.GetArch(ir.Architecture)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Unsupported architecture '%s' for distribution '%s'", ir.Architecture, request.Distribution), http.StatusBadRequest)
+			return
+		}
+		imageType, err := arch.GetImageType(ir.ImageType)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Unsupported image type '%s' for %s/%s", ir.ImageType, ir.Architecture, request.Distribution), http.StatusBadRequest)
+			return
+		}
+		repositories := make([]rpmmd.RepoConfig, len(ir.Repositories))
+		for j, repo := range ir.Repositories {
+			repositories[j].BaseURL = repo.Baseurl
+			repositories[j].RHSM = true
```

Comment:
I want to make this part better in a followup, we still wanna be able to build (non-rhel) images on systems that don't have rhsm credentials in place.

Reply:
This API is only meant for making official RHEL images (for now), so it's good to go in as is, right?

Reply:
Right but it might be a pain where you need a registered system if you want to run this locally. For local development it's so much easier to build fedora images :/